Python: pipenv update (adds atomicwrites)

Simran-B opened the pull request:
I followed the [Python readme](https://github.com/AcademySoftwareFoundation/aswf-docker/blob/master/python/README.md#for-developers) but ran into below problem with a fresh virtualenv:

```
aswf-docker> pipenv shell

Creating a virtualenv for this project...
Pipfile: C:\Daten\Olive\Docker\aswf-docker\Pipfile
Using C:/Python37/python.exe (3.7.9) to create virtualenv...
[=   ] Creating virtual environment...created virtual environment CPython3.7.9.final.0-64 in 2746ms
  creator CPython3Windows(dest=C:\Users\Simran\.virtualenvs\aswf-docker-_K9OxNZo, clear=False, no_vcs_ignore=False, global=False)
  seeder FromAppData(download=False, pip=bundle, setuptools=bundle, wheel=bundle, via=copy, app_data_dir=C:\Users\Simran\AppData\Local\pypa\virtualenv)
    added seed packages: pip==20.2.4, setuptools==50.3.2, wheel==0.35.1
  activators BashActivator,BatchActivator,FishActivator,PowerShellActivator,PythonActivator,XonshActivator

Successfully created virtual environment!
Virtualenv location: C:\Users\Simran\.virtualenvs\aswf-docker-_K9OxNZo
Launching subshell in virtual environment...
Microsoft Windows [Version 10.0.19041.630]
(c) 2020 Microsoft Corporation. Alle Rechte vorbehalten.


aswf-docker> pipenv install --dev

Installing dependencies from Pipfile.lock (81b6fd)...
  ================================ 76/76 - 00:01:03


aswf-docker> pytest python/aswfdocker

Traceback (most recent call last):
  File "c:\python37\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "c:\python37\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "C:\Users\Simran\.virtualenvs\aswf-docker-_K9OxNZo\Scripts\pytest.exe\__main__.py", line 4, in <module>
  File "c:\users\simran\.virtualenvs\aswf-docker-_k9oxnzo\lib\site-packages\pytest\__init__.py", line 5, in <module>
    from _pytest.assertion import register_assert_rewrite
  File "c:\users\simran\.virtualenvs\aswf-docker-_k9oxnzo\lib\site-packages\_pytest\assertion\__init__.py", line 8, in <module>
    from _pytest.assertion import rewrite
  File "c:\users\simran\.virtualenvs\aswf-docker-_k9oxnzo\lib\site-packages\_pytest\assertion\rewrite.py", line 300, in <module>
    from atomicwrites import atomic_write
ModuleNotFoundError: No module named 'atomicwrites'
```

Updating the dependencies with `pipenv update` added the required atomicwrites module.

Simran-B commented on Pipfile.lock:
```diff
@@ -257,7 +267,7 @@
                 "sha256:9f47eda37229f68eee03b24b9748937c7dc3868f906e8ba69fbcbdd3bc5dc3e2"
             ],
-            "markers": "python_version >= '2.7' and python_version not in '3.0, 3.1, 3.2, 3.3, 3.4'",
+            "markers": "sys_platform == 'win32' and sys_platform == 'win32'",
```
This one is strange... the same condition twice? Does updating on another OS produce something different here?

jfpanisset reviewed (approved):
I don't understand enough about pipenv to answer the questions, but it looks like this would be a good thing. Looks like you are trying to build aswf-container on win32, how far along have you gotten so far? Back when we were using Azure Pipelines, it was possible to "build inside a container" on Windows, hopefully at some point this support will come to GitHub Actions.

Simran-B commented:
@jfpanisset I wanted to make sure that the Qt image actually builds and works on my machine before I change the Qt version to 5.15.2 (#98), but I ran into the problem with the `atomicwrites` module being missing (added by this PR) and a few assertion errors when running `pytest` (#95). With the latter two fixes, `aswfdocker --verbose build -t IMAGE --group base --version 2021 --target baseqt`, `pytest python/aswfdocker` etc. complete successfully on my Windows 10 machine using Docker with Linux containers (WSL2 backend).

<details><summary>Full log</summary><pre><code>PS C:\Daten\Olive\Docker\aswf-docker> pipenv shell
Launching subshell in virtual environment...
PowerShell 7.1.0
Copyright (c) Microsoft Corporation.

https://aka.ms/powershell
Type 'help' to get help.

PS C:\Daten\Olive\Docker\aswf-docker> pipenv install --dev
Installing dependencies from Pipfile.lock (81b6fd)...
  ================================ 10/10 - 00:00:20
PS C:\Daten\Olive\Docker\aswf-docker> aswfdocker --verbose build -t IMAGE --group base --version 2021 --target baseqt
DEBUG:aswfdocker.groupinfo:GroupInfo: type=ImageType.IMAGE names=('base',) versions=['2021'] images=['base', 'baseqt'] targets=('baseqt',)
DEBUG:aswfdocker.groupinfo:Skipping target base
DEBUG:aswfdocker.groupinfo:iter image=baseqt
DEBUG:aswfdocker.groupinfo:iter version=2021
DEBUG:aswfdocker.builder:Repo root: C:\Daten\Olive\Docker\aswf-docker
INFO:aswfdocker.builder:Building: 'docker buildx bake -f C:\\Users\\Simran\\AppData\\Local\\Temp\\docker-bake-IMAGE-base-2021.json --progress auto'
[+] Building 34.5s (37/37) FINISHED
 => [internal] load build definition from Dockerfile                                                                                                                                                        0.1s
 => => transferring dockerfile: 5.09kB                                                                                                                                                                      0.0s
 => [internal] load .dockerignore                                                                                                                                                                           0.1s
 => => transferring context: 2B                                                                                                                                                                             0.0s
 => [internal] load metadata for docker.io/aswftesting/ci-package-glfw:2021-3.1.2                                                                                                                           2.2s
 => [internal] load metadata for docker.io/aswftesting/ci-package-tbb:2021-2019_U6                                                                                                                          2.3s
 => [internal] load metadata for docker.io/aswftesting/ci-package-log4cplus:2021-1.1.2                                                                                                                      2.2s
 => [internal] load metadata for docker.io/aswftesting/ci-package-glew:2021-2.1.0                                                                                                                           1.7s
 => [internal] load metadata for docker.io/aswftesting/ci-package-python:2021-3.7.3                                                                                                                         1.6s
 => [internal] load metadata for docker.io/aswftesting/ci-package-pyside:2021-5.12.6                                                                                                                        3.1s
 => [internal] load metadata for docker.io/aswftesting/ci-package-boost:2021-1.70.0                                                                                                                         1.9s
 => [internal] load metadata for docker.io/aswftesting/ci-package-cmake:2021-3.18.4                                                                                                                         2.1s
 => [internal] load metadata for docker.io/aswftesting/ci-package-cppunit:2021-1.15.1                                                                                                                       2.1s
 => [internal] load metadata for docker.io/aswftesting/ci-common:2-clang10                                                                                                                                  0.0s
 => [internal] load metadata for docker.io/aswftesting/ci-package-qt:2021-5.12.8                                                                                                                            0.0s
 => [internal] load build context                                                                                                                                                                           0.1s
 => => transferring context: 2.52kB                                                                                                                                                                         0.0s
 => [ci-package-python 1/1] FROM docker.io/aswftesting/ci-package-python:2021-3.7.3@sha256:3b957bb98d392db49445bc8b0fde5309286b90b26b2d3f21d654beba1d663730                                                23.6s
 => => resolve docker.io/aswftesting/ci-package-python:2021-3.7.3@sha256:3b957bb98d392db49445bc8b0fde5309286b90b26b2d3f21d654beba1d663730                                                                   0.0s
 => => sha256:930704de2ce0c77882c9b82da962bfb627aeddd0ca6e48d43a61504b81a48887 69.01MB / 69.01MB                                                                                                           20.2s
 => => sha256:3b957bb98d392db49445bc8b0fde5309286b90b26b2d3f21d654beba1d663730 523B / 523B                                                                                                                  0.0s
 => => sha256:45d6528c101cc4e58a3ae0e4167176115f4301ef97e8c624dffafd9258d6fab2 4.79kB / 4.79kB                                                                                                              0.0s
 => => extracting sha256:930704de2ce0c77882c9b82da962bfb627aeddd0ca6e48d43a61504b81a48887                                                                                                                   2.9s
 => [ci-package-boost 1/1] FROM docker.io/aswftesting/ci-package-boost:2021-1.70.0@sha256:5705dcc0719338f698f84de19ab6fbc7aad03677d35e0f390310b6d47e948cc5                                                 17.2s
 => => resolve docker.io/aswftesting/ci-package-boost:2021-1.70.0@sha256:5705dcc0719338f698f84de19ab6fbc7aad03677d35e0f390310b6d47e948cc5                                                                   0.0s
 => => sha256:5705dcc0719338f698f84de19ab6fbc7aad03677d35e0f390310b6d47e948cc5 523B / 523B                                                                                                                  0.0s
 => => sha256:c29ebf73428fa1ff481ef7b771941e050b52c457b319c23b2ebb90b274183d4f 4.31kB / 4.31kB                                                                                                              0.0s
 => => sha256:079acd0735c20dde178c039622935c4893f275a6897bfc49b5a8d10ab3a4e06a 17.53MB / 17.53MB                                                                                                           13.5s
 => => extracting sha256:079acd0735c20dde178c039622935c4893f275a6897bfc49b5a8d10ab3a4e06a                                                                                                                   3.1s
 => [ci-package-qt 1/1] FROM docker.io/aswftesting/ci-package-qt:2021-5.12.8                                                                                                                                0.0s
 => => resolve docker.io/aswftesting/ci-package-qt:2021-5.12.8                                                                                                                                              0.0s
 => [ci-package-glfw 1/1] FROM docker.io/aswftesting/ci-package-glfw:2021-3.1.2@sha256:c5a502d990c0028e96bb4c426325c9eb9bfd85ac6b56e7704ca617be6d88434f                                                     1.3s
 => => resolve docker.io/aswftesting/ci-package-glfw:2021-3.1.2@sha256:c5a502d990c0028e96bb4c426325c9eb9bfd85ac6b56e7704ca617be6d88434f                                                                     0.0s
 => => sha256:c5a502d990c0028e96bb4c426325c9eb9bfd85ac6b56e7704ca617be6d88434f 520B / 520B                                                                                                                  0.0s
 => => sha256:21bc9b27275f7a58e6c2477e2e6f68c7431796cf50ddae8518d494187920fddf 4.32kB / 4.32kB                                                                                                              0.0s
 => => sha256:77b544080824e16948a4cfad8431ed859a2cb02577b846023ae0aca39e0cad4d 77.53kB / 77.53kB                                                                                                            1.0s
 => => extracting sha256:77b544080824e16948a4cfad8431ed859a2cb02577b846023ae0aca39e0cad4d                                                                                                                   0.0s
 => [ci-package-log4cplus 1/1] FROM docker.io/aswftesting/ci-package-log4cplus:2021-1.1.2@sha256:d34acd118bb28a097c7b49f94473e55e282816859497433c76e35b57e4c90370                                          15.6s
 => => resolve docker.io/aswftesting/ci-package-log4cplus:2021-1.1.2@sha256:d34acd118bb28a097c7b49f94473e55e282816859497433c76e35b57e4c90370                                                                0.0s
 => => sha256:d34acd118bb28a097c7b49f94473e55e282816859497433c76e35b57e4c90370 521B / 521B                                                                                                                  0.0s
 => => sha256:ef66749acd29d287465f0b351daad7cdf0d13789bfd63522b8a9a45b3639a887 4.45kB / 4.45kB                                                                                                              0.0s
 => => sha256:4c64d136159d8a22c3a32cbdafa75fcdffe9b215b1bcc0d55d514c823bf2711f 442.80kB / 442.80kB                                                                                                         15.0s
 => => extracting sha256:4c64d136159d8a22c3a32cbdafa75fcdffe9b215b1bcc0d55d514c823bf2711f                                                                                                                   0.1s
 => [ci-package-tbb 1/1] FROM docker.io/aswftesting/ci-package-tbb:2021-2019_U6@sha256:c55042385e6cea7aa1e308980a65b906e272aa62480a8f531d575118071a4af7                                                     4.3s
 => => resolve docker.io/aswftesting/ci-package-tbb:2021-2019_U6@sha256:c55042385e6cea7aa1e308980a65b906e272aa62480a8f531d575118071a4af7                                                                    0.0s
 => => sha256:c55042385e6cea7aa1e308980a65b906e272aa62480a8f531d575118071a4af7 522B / 522B                                                                                                                  0.0s
 => => sha256:b2740d7b754bcac44eb000bda01e994e9ee1285399aab0637c9e0bf43e474b5a 4.37kB / 4.37kB                                                                                                              0.0s
 => => sha256:671eadaece23f36339f8485fe253788e9e3abb13acd08f97df9e8b649d01cf10 4.15MB / 4.15MB                                                                                                              3.6s
 => => extracting sha256:671eadaece23f36339f8485fe253788e9e3abb13acd08f97df9e8b649d01cf10                                                                                                                   0.2s
 => [ci-package-glew 1/1] FROM docker.io/aswftesting/ci-package-glew:2021-2.1.0@sha256:23a2ce7c776a092100340e6bce4304a6af0f0d69fa0f458669c83f220dee745d                                                    15.6s
 => => resolve docker.io/aswftesting/ci-package-glew:2021-2.1.0@sha256:23a2ce7c776a092100340e6bce4304a6af0f0d69fa0f458669c83f220dee745d                                                                     0.0s
 => => sha256:fc7229471ed003531c9d6bd2fe20997b2538145a1d4b662c36feb1a2546a49f2 4.36kB / 4.36kB                                                                                                              0.0s
 => => sha256:23a2ce7c776a092100340e6bce4304a6af0f0d69fa0f458669c83f220dee745d 521B / 521B                                                                                                                  0.0s
 => => sha256:668e3fc064ac080b6482168ea565dcbe63d995603b30f2510b6ce17e4c9cd20d 793.87kB / 793.87kB                                                                                                         14.6s
 => => extracting sha256:668e3fc064ac080b6482168ea565dcbe63d995603b30f2510b6ce17e4c9cd20d                                                                                                                   0.1s
 => [ci-package-cppunit 1/1] FROM docker.io/aswftesting/ci-package-cppunit:2021-1.15.1@sha256:7e4b9983be2a5ca63eb7f18d9f0c505dab478a894c46ab997a7827e311140a8c                                              4.2s
 => => resolve docker.io/aswftesting/ci-package-cppunit:2021-1.15.1@sha256:7e4b9983be2a5ca63eb7f18d9f0c505dab478a894c46ab997a7827e311140a8c                                                                 0.0s
 => => sha256:7e4b9983be2a5ca63eb7f18d9f0c505dab478a894c46ab997a7827e311140a8c 522B / 522B                                                                                                                  0.0s
 => => sha256:c86d66a1fb0c3b65828f90efb6410dfc4ebbf3160290082b2be518a1e5260b93 4.31kB / 4.31kB                                                                                                              0.0s
 => => sha256:ce9e1bbfd4f931be9a724a49109ba9ceffbbd84d957f3e88f50afb4fc42db988 5.29MB / 5.29MB                                                                                                              3.5s
 => => extracting sha256:ce9e1bbfd4f931be9a724a49109ba9ceffbbd84d957f3e88f50afb4fc42db988                                                                                                                   0.4s
 => CACHED [ci-package-cmake 1/1] FROM docker.io/aswftesting/ci-package-cmake:2021-3.18.4@sha256:1cb0dc38e9ffbfd619fbdd55d75c2327a26dadaa5dc8ad169c0f3680af3b2463                                           0.0s
 => [ci-package-pyside 1/1] FROM docker.io/aswftesting/ci-package-pyside:2021-5.12.6@sha256:b239d5e31f925939e03a9f77bfff6e26b09a9b4d11d638b6f44fdf5922e0a0fc                                               14.7s
 => => resolve docker.io/aswftesting/ci-package-pyside:2021-5.12.6@sha256:b239d5e31f925939e03a9f77bfff6e26b09a9b4d11d638b6f44fdf5922e0a0fc                                                                  0.0s
 => => sha256:b239d5e31f925939e03a9f77bfff6e26b09a9b4d11d638b6f44fdf5922e0a0fc 523B / 523B                                                                                                                  0.0s
 => => sha256:32d51873851894bc0009abdffadad6e3a7c1fd1a2955494fa20e66e4486c4b4a 4.37kB / 4.37kB                                                                                                              0.0s
 => => sha256:95e40c24f02f2f964c51146314cf5088ef2d9c4b624e3c7d719872712843f003 18.04MB / 18.04MB                                                                                                           13.4s
 => => extracting sha256:95e40c24f02f2f964c51146314cf5088ef2d9c4b624e3c7d719872712843f003                                                                                                                   0.8s
 => CACHED [ci-baseqt 1/12] FROM docker.io/aswftesting/ci-common:2-clang10                                                                                                                                  0.0s
 => [ci-baseqt 2/12] COPY --from=ci-package-boost /. /usr/local/                                                                                                                                            1.7s
 => [ci-baseqt 3/12] COPY --from=ci-package-cmake /. /usr/local/                                                                                                                                            3.0s
 => [ci-baseqt 4/12] COPY --from=ci-package-cppunit /. /usr/local/                                                                                                                                          0.3s
 => [ci-baseqt 5/12] COPY --from=ci-package-glew /. /usr/local/                                                                                                                                             0.1s
 => [ci-baseqt 6/12] COPY --from=ci-package-glfw /. /usr/local/                                                                                                                                             0.1s
 => [ci-baseqt 7/12] COPY --from=ci-package-log4cplus /. /usr/local/                                                                                                                                        0.1s
 => [ci-baseqt 8/12] COPY --from=ci-package-python /. /usr/local/                                                                                                                                           1.3s
 => [ci-baseqt 9/12] COPY --from=ci-package-tbb /. /usr/local/                                                                                                                                              0.1s
 => [ci-baseqt 10/12] COPY --from=ci-package-qt /. /usr/local/                                                                                                                                              2.2s
 => [ci-baseqt 11/12] COPY --from=ci-package-pyside /. /usr/local/                                                                                                                                          0.4s
 => [ci-baseqt 12/12] COPY ci-baseqt/README.md ci-baseqt/image.yaml /usr/local/aswf/                                                                                                                        0.0s
 => exporting to image                                                                                                                                                                                      2.8s
 => => exporting layers                                                                                                                                                                                     2.6s
 => => writing image sha256:eeeb3bcdd5c955e575c9b722e297965688dc064346e2ae858c876873dea12126                                                                                                                0.0s
 => => naming to docker.io/aswftesting/ci-baseqt:2021                                                                                                                                                       0.0s
 => => naming to docker.io/aswftesting/ci-baseqt:2021.0                                                                                                                                                     0.0s
 => => naming to docker.io/aswftesting/ci-baseqt:preview

PS C:\Daten\Olive\Docker\aswf-docker> black python
All done! ✨ 🍰 ✨
23 files left unchanged.

PS C:\Daten\Olive\Docker\aswf-docker> pytest python/aswfdocker
============================================================================================= test session starts ==============================================================================================
platform win32 -- Python 3.7.9, pytest-6.1.2, py-1.9.0, pluggy-0.13.1
rootdir: C:\Daten\Olive\Docker\aswf-docker
plugins: cov-2.10.1
collected 29 items

python\aswfdocker\tests\test_builder.py .......                                                                                                                                                           [ 24%]
python\aswfdocker\tests\test_dockergen.py ..                                                                                                                                                              [ 31%]
python\aswfdocker\tests\test_index.py ....                                                                                                                                                                [ 44%]
python\aswfdocker\tests\test_migrater.py ....                                                                                                                                                             [ 58%]
python\aswfdocker\tests\test_releaser.py ..                                                                                                                                                               [ 65%]
python\aswfdocker\tests\test_settings.py .                                                                                                                                                                [ 68%]
python\aswfdocker\tests\test_utils.py .........                                                                                                                                                           [100%]

============================================================================================= 29 passed in 11.57s ==============================================================================================

PS C:\Daten\Olive\Docker\aswf-docker> mypy python/aswfdocker
Success: no issues found in 23 source files

PS C:\Daten\Olive\Docker\aswf-docker> prospector -F python/aswfdocker
Check Information
=================
         Started: 2020-12-07 11:36:59.664084
        Finished: 2020-12-07 11:39:18.012196
      Time Taken: 138.35 seconds
       Formatter: grouped
        Profiles: default, full_pep8, no_doc_warnings, no_test_warnings, strictness_medium, strictness_high, strictness_veryhigh, no_member_warnings
      Strictness: None
  Libraries Used:
       Tools Run: dodgy, mccabe, pep8, profile-validator, pyflakes, pylint
  Messages Found: 0</code></pre></details>

If I switch to Windows containers, buildx fails right away because of the mount type:

```
PS C:\Daten\Olive\Docker\aswf-docker> aswfdocker --verbose build -t IMAGE --group base --version 2021 --target baseqt
DEBUG:aswfdocker.groupinfo:GroupInfo: type=ImageType.IMAGE names=('base',) versions=['2021'] images=['base', 'baseqt'] targets=('baseqt',)
DEBUG:aswfdocker.groupinfo:Skipping target base
DEBUG:aswfdocker.groupinfo:iter image=baseqt
DEBUG:aswfdocker.groupinfo:iter version=2021
DEBUG:aswfdocker.builder:Repo root: C:\Daten\Olive\Docker\aswf-docker
INFO:aswfdocker.builder:Building: 'docker buildx bake -f C:\\Users\\Simran\\AppData\\Local\\Temp\\docker-bake-IMAGE-base-2021.json --progress auto'
[+] Building 0.1s (2/2) FINISHED
 => [internal] load .dockerignore                                                                                  0.1s
 => => transferring context: 2B                                                                                    0.0s
 => [internal] load build definition from Dockerfile                                                               0.1s
 => => transferring dockerfile: 5.09kB                                                                             0.0s
failed to solve: rpc error: code = Unknown desc = failed to solve with frontend dockerfile.v0: failed to read dockerfile: failed to mount C:\ProgramData\Docker\tmp\buildkit-mount199307067: [{Type:bind Source:C:\ProgramData\Docker\windowsfilter\j5rqylu7cnyb2aay8kv92ld7v Options:[rbind ro]}]: invalid windows mount type: 'bind'
Traceback (most recent call last):
  File "C:\Users\Simran\.virtualenvs\aswf-docker-_K9OxNZo\Scripts\aswfdocker-script.py", line 33, in <module>
    sys.exit(load_entry_point('aswfdocker', 'console_scripts', 'aswfdocker')())
  File "C:\Users\Simran\.virtualenvs\aswf-docker-_K9OxNZo\lib\site-packages\click\core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "C:\Users\Simran\.virtualenvs\aswf-docker-_K9OxNZo\lib\site-packages\click\core.py", line 782, in main
    rv = self.invoke(ctx)
  File "C:\Users\Simran\.virtualenvs\aswf-docker-_K9OxNZo\lib\site-packages\click\core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "C:\Users\Simran\.virtualenvs\aswf-docker-_K9OxNZo\lib\site-packages\click\core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "C:\Users\Simran\.virtualenvs\aswf-docker-_K9OxNZo\lib\site-packages\click\core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "C:\Users\Simran\.virtualenvs\aswf-docker-_K9OxNZo\lib\site-packages\click\decorators.py", line 73, in new_func
    return ctx.invoke(f, obj, *args, **kwargs)
  File "C:\Users\Simran\.virtualenvs\aswf-docker-_K9OxNZo\lib\site-packages\click\core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "c:\daten\olive\docker\aswf-docker\python\aswfdocker\cli\aswfdocker.py", line 172, in build
    b.build(dry_run=dry_run, progress=progress)
  File "c:\daten\olive\docker\aswf-docker\python\aswfdocker\builder.py", line 95, in build
    subprocess.run(cmd, shell=True, check=True, cwd=self.build_info.repo_root)
  File "c:\python37\lib\subprocess.py", line 512, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command 'docker buildx bake -f C:\Users\Simran\AppData\Local\Temp\docker-bake-IMAGE-base-2021.json --progress auto' returned non-zero exit status 1.
```

The actual problem here seems to be that BuildKit is "[Only supported for building Linux containers](https://docs.docker.com/develop/develop-images/build_enhancements/#limitations)" however. Someone [worked on Windows support](https://github.com/moby/buildkit/issues/616) but there a several pending PRs and upstream TODOs, so it's not landing anytime soon (if at all).

---

GitHub Actions only supports Linux containers unfortunately, but who knows what the future brings. A huge downside of Windows containers is their size however. If a blank Windows Server Core is sufficient then it's fine, as GitHub caches this base image, but if you need Visual Studio in that container then it will be a few gigabytes in size, possibly exceeding the space quota of free GitHub Actions runners. I also ran into profound issues with CMake's self-check and Visual Studio's PDB server which doesn't seem to work with Hyper-V virtualization (I suppose this isn't an issue if process isolation is used instead, but then the guest OS is no longer independent of the host OS - i.e. they must match in type and build number). Overall, my WCOW experience was very frustrating so far. I don't plan on experimenting with it any further, especially because of GitHub Actions' limitations. We will probably use Conan and the GitHub-hosted runners directly for our project (there's no strict requirement for reproducible builds, so no need for a fully controlled environment).

aloysbaillet commented:
Thanks for the information! The "bind" instructions are just to speed up rebuilding so that downloaded files are kept around, you could remove them and try again if you feel adventurous...

Simran-B commented:
Do you refer to these mount instructions?
```Dockerfile
RUN --mount=type=cache,target=/tmp/ccache \
    --mount=type=cache,target=/tmp/downloads \
```

They are of type _cache_, not _bind_. I can't find any mounts of type _bind_ that I could change. I removed all _cache_  mounts and added `--no-cache` to the buildx commands, but that doesn't change anything. It still tries to bind something - I guess this is an internal BuildKit thing that cannot be turned off?

TBBle commented:
Yeah, that's an internal attempt to bind-mount the directory the Dockerfile was copied to, I think. It's one of the things that the "several pending PRs and upstream TODOs" mentioned above resolve, that're blocking BuildKit on Windows.

aloysbaillet commented:
So I opened #107 to fix the tests and the missing atomicwrites on Windows, that said buildx will still not work on Windows...
Happy to open a new issue to track windows support for building the linux docker packages if you want.
